services/oauth/v2/http/transport.go

```diff
@@ -0,0 +1,276 @@
+	t := &OAuthTransport{
+		oauthHandler:         *args.OAuthHandler,
+		Augmenter:            args.Augmenter,
+		Transport:            args.OriginalTransport,
+		log:                  args.logger,
+		flow:                 args.FlowType,
+		getAuthErrorCategory: args.GetAuthErrorCategory,
+		stats:                args.stats,
+	}
+	if t.log == nil {
+		t.log = logger.NewLogger()
+	}
+	if t.stats == nil {
+		t.stats = stats.Default
+	}
+	t.log = t.log.Child("OAuthHttpTransport")
+	return t
+}
+
+func (t *OAuthTransport) preRoundTrip(rts *roundTripState) *http.Response {
+	if t.Augmenter == nil {
+		return nil
+	}
+	body, err := io.ReadAll(rts.req.Body)
+	if err != nil {
+		t.log.Errorn("failed to read request body",
+			obskit.DestinationID(rts.destination.ID),
+			obskit.WorkspaceID(rts.destination.WorkspaceID),
+			obskit.DestinationType(rts.destination.DefinitionName),
+			logger.NewStringField("flow", string(t.flow)))
+		return httpResponseCreator(http.StatusInternalServerError, []byte(fmt.Errorf("failed to read request body pre roundTrip: %w", err).Error()))
+	}
+	statusCode, authResponse, err := t.oauthHandler.FetchToken(rts.refreshTokenParams)
+	if statusCode == http.StatusOK {
+		rts.req = rts.req.WithContext(cntx.CtxWithSecret(rts.req.Context(), authResponse.Account.Secret))
+		err = t.Augmenter.Augment(rts.req, body, authResponse.Account.Secret)
+		if err != nil {
+			t.log.Debugn("failed to augment the secret",
+				logger.NewErrorField(err))
+			return httpResponseCreator(http.StatusInternalServerError, []byte(fmt.Errorf("failed to augment the secret pre roundTrip: %w", err).Error()))
+		}
+		return nil
+	} else if authResponse != nil && authResponse.Err == common.RefTokenInvalidGrant {
+		t.oauthHandler.AuthStatusToggle(&oauth.AuthStatusToggleParams{
+			Destination:     rts.destination,
+			WorkspaceID:     rts.destination.WorkspaceID,
+			RudderAccountID: rts.accountID,
+			StatPrefix:      common.AuthStatusInActive,
+			AuthStatus:      common.CategoryAuthStatusInactive,
+		})
+		return httpResponseCreator(http.StatusBadRequest, []byte(err.Error()))
+	}
+	return httpResponseCreator(statusCode, []byte(err.Error()))
+}
+
+func (t *OAuthTransport) postRoundTrip(rts *roundTripState) (*http.Response, error) {
+	respData, err := io.ReadAll(rts.res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body post RoundTrip: %w", err)
+	}
+	interceptorResp := oauth.OAuthInterceptorResponse{}
+	// internal function
+	applyInterceptorRespToHttpResp := func() {
+		var interceptorRespBytes []byte
+		transResp := oauth.TransportResponse{
+			OriginalResponse:    string(respData),
+			InterceptorResponse: interceptorResp,
+		}
+		interceptorRespBytes, _ = json.Marshal(transResp)
+		rts.res.Body = io.NopCloser(bytes.NewReader(interceptorRespBytes))
+	}
+	authErrorCategory, err := t.getAuthErrorCategory(respData)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get auth error category: %s", string(respData))
```

Comment:
[nit-pik] Avoid using the word fail, or error instead the error messages. When an error is wrapped multiple type it makes the final message more readable.

```suggestion
		return nil, fmt.Errorf("getting auth error category: %s", string(respData))
```